pkg/controller/tas/node_failure_controller.go

```diff
@@ -0,0 +1,202 @@
+		clock:  clock.RealClock{},
+	}
+}
+
+func (r *nodeFailureReconciler) SetupWithManager(mgr ctrl.Manager, cfg *config.Configuration) (string, error) {
+	return NodeFailureControllerName, builder.ControllerManagedBy(mgr).
+		Named(NodeFailureControllerName).
+		For(&corev1.Node{}, builder.WithPredicates(predicate.Funcs{
+			CreateFunc:  func(e event.CreateEvent) bool { return false },
+			UpdateFunc:  func(e event.UpdateEvent) bool { return true },
+			DeleteFunc:  func(e event.DeleteEvent) bool { return true },
+			GenericFunc: func(e event.GenericEvent) bool { return false },
+		})).
+		Complete(r)
+}
+
+func (r *nodeFailureReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx).WithValues("node", req.Name)
+	ctx = ctrl.LoggerInto(ctx, log)
+
+	log.V(3).Info("Getting node", "name", req.Name, "namespacename", req.NamespacedName)
+	var node corev1.Node
+	err := r.client.Get(ctx, req.NamespacedName, &node)
+	nodeExists := !apierrors.IsNotFound(err)
+	if err != nil && nodeExists {
+		log.Error(err, "Failed to get Node")
+		return ctrl.Result{}, err
+	}
+
+	nodeIsReady := false
+	markFailed := false
+	var readyTransitionTime *metav1.Time
+	var readyStatus corev1.ConditionStatus
+	requeueAfter := time.Duration(0)
+	if nodeExists {
+		readyStatus, readyTransitionTime = r.getNodeConditionState(&node)
+		nodeIsReady = (readyStatus == corev1.ConditionTrue)
+		log.V(3).Info("Reconciling node", "found", nodeExists, "ready", nodeIsReady)
+		if nodeIsReady {
+			return ctrl.Result{}, nil
+		}
+		if readyTransitionTime != nil {
+			timeSinceNotReady := r.clock.Now().Sub(readyTransitionTime.Time)
+			if timeSinceNotReady >= NodeFailureDelay {
+				markFailed = true
+			} else {
+				requeueAfter = NodeFailureDelay - timeSinceNotReady
+			}
+		} else { // Node exists, not ready, but no transition time? Treat as failed immediately.
+			markFailed = true
+		}
+
+	} else {
+		log.V(3).Info("Node not found, assuming deleted")
+		markFailed = true
+	}
+
+	if markFailed {
+		updateErr := r.updateWorkloadsForFailedNode(ctx, req.Name)
+		if updateErr != nil {
+			return ctrl.Result{}, updateErr
+		}
+		return ctrl.Result{}, nil
+	}
+	return ctrl.Result{RequeueAfter: requeueAfter}, nil
+}
+
+// updateWorkloadsForNode finds workloads with pods on the specified node
+// and updates their status based on whether the node is currently considered failed.
+func (r *nodeFailureReconciler) updateWorkloadsForFailedNode(ctx context.Context, nodeName string) error {
+	log := ctrl.LoggerFrom(ctx)
+	var podList corev1.PodList
+	if err := r.client.List(ctx, &podList, client.MatchingFields{indexer.PodNodeNameIndexKey: nodeName}); err != nil {
+		log.Error(err, "Failed to list pods on node", "nodeName", nodeName)
+		return err
+	}
+	consideredWorkloads := sets.New[types.NamespacedName]()
+	var workloadPatchErrors []error
+
+	for i := range podList.Items {
+		pod := &podList.Items[i]
```

Comment:
```suggestion
	for _, pod  := range podList.Items {
```
Nit, wdyt about such a refactoring for it to be a little more conventional